p2p/p2p_test.go

```diff
@@ -380,8 +384,116 @@ func TestGossiping(t *testing.T) {
 		if !reflect.DeepEqual(bmEnc, res) {
 			t.Fatalf("Didn't receive the correct message")
 		}
-	case <-time.After(10 * time.Second):
+	case <-time.After(30 * time.Second):
 		t.Fatalf("Did not receive message from %s", nodeB.host.hostAddr)
 	}
 
 }
+
+func TestP2pReceiveChan(t *testing.T) {
+	//Create nodeA
+	testServiceConfigA := &Config{
+		BootstrapNodes: nil,
+		NoBootstrap:    true,
+		NoMdns:         true,
+		Port:           7000,
+		RandSeed:       1,
+	}
+
+	msgRecChan := make(chan BlockAnnounceMessage)
+
+	nodeA := startNewService(t, testServiceConfigA, nil, msgRecChan)
+	defer nodeA.Stop()
+
+	// Create nodeB
+	nodeAAddr := nodeA.host.fullAddrs()[0]
+	testServiceConfigB := &Config{
+		BootstrapNodes: []string{
+			nodeAAddr.String(),
+		},
+		Port:     7001,
+		NoMdns:   true,
+		RandSeed: 2,
+	}
+	msgSendChan := make(chan []byte)
+	nodeB := startNewService(t, testServiceConfigB, msgSendChan, nil)
+	defer nodeB.Stop()
+
+	// Create & set a custom stream handler for node B
+	nodeB.host.registerStreamHandler(func(stream net.Stream) {
```

Comment:
Lets avoid all this code reuse by refactoring `Service.handleStream`. The majority of the function is parsing the message, lets move that logic to another method on service, and then call that here.